google/cloud/storage/internal/hash_function_impl_test.cc

```diff
@@ -35,82 +38,191 @@ auto constexpr kEmptyStringMD5Hash = "1B2M2Y8AsgTpgAmY7PhCfg==";
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.substr(npos, 5);
+    EXPECT_STATUS_OK(function.Update(npos, message));
+    EXPECT_THAT(function.Update(npos, message),
+                StatusIs(StatusCode::kInvalidArgument));
+  }
+  auto const actual = function.Finish();
+  EXPECT_THAT(actual.crc32c, IsEmpty());
+  EXPECT_THAT(actual.md5, kQuickFoxMD5Hash);
+
+  auto const a2 = function.Finish();
+  EXPECT_THAT(a2.crc32c, IsEmpty());
+  EXPECT_THAT(a2.md5, kQuickFoxMD5Hash);
+}
+
+TEST(HashFunctionImplTest, MD5StringViewWithCrc) {
+  MD5HashFunction function;
+  auto payload = absl::string_view{kQuickFox};
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.substr(npos, 5);
+    auto const message_crc = storage_internal::Crc32c(message);
+    EXPECT_STATUS_OK(function.Update(npos, message, message_crc));
+    EXPECT_THAT(function.Update(npos, message, message_crc),
+                StatusIs(StatusCode::kInvalidArgument));
+  }
+  auto const actual = function.Finish();
+  EXPECT_THAT(actual.crc32c, IsEmpty());
+  EXPECT_THAT(actual.md5, kQuickFoxMD5Hash);
+
+  auto const a2 = function.Finish();
+  EXPECT_THAT(a2.crc32c, IsEmpty());
+  EXPECT_THAT(a2.md5, kQuickFoxMD5Hash);
+}
+
+TEST(HashFunctionImplTest, MD5Cord) {
+  MD5HashFunction function;
+  auto payload = absl::Cord(absl::string_view{kQuickFox});
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.Subcord(npos, 5);
+    auto const message_crc = storage_internal::Crc32c(message);
+    EXPECT_STATUS_OK(function.Update(npos, message, message_crc));
+    EXPECT_THAT(function.Update(npos, message, message_crc),
+                StatusIs(StatusCode::kInvalidArgument));
+  }
+  auto const actual = function.Finish();
+  EXPECT_THAT(actual.crc32c, IsEmpty());
+  EXPECT_THAT(actual.md5, kQuickFoxMD5Hash);
+
+  auto const a2 = function.Finish();
+  EXPECT_THAT(a2.crc32c, IsEmpty());
+  EXPECT_THAT(a2.md5, kQuickFoxMD5Hash);
+}
+
+TEST(HashFunctionImplTest, CompositeEmpty) {
   CompositeFunction function(absl::make_unique<MD5HashFunction>(),
                              absl::make_unique<Crc32cHashFunction>());
   auto result = std::move(function).Finish();
   EXPECT_THAT(result.crc32c, kEmptyStringCrc32cChecksum);
   EXPECT_THAT(result.md5, kEmptyStringMD5Hash);
 }
 
-TEST(HasFunctionImplTest, QuickComposite) {
+TEST(HashFunctionImplTest, CompositeQuick) {
   CompositeFunction function(absl::make_unique<MD5HashFunction>(),
                              absl::make_unique<Crc32cHashFunction>());
-  Update(function, "The quick");
-  Update(function, " brown");
-  Update(function, " fox jumps over the lazy dog");
+  function.Update("The quick");
+  function.Update(" brown");
+  function.Update(" fox jumps over the lazy dog");
   auto result = std::move(function).Finish();
   EXPECT_THAT(result.crc32c, kQuickFoxCrc32cChecksum);
   EXPECT_THAT(result.md5, kQuickFoxMD5Hash);
```

Comment:
should we call `Finish()` twice and see if `ubsan` hollers?

Reply:
Done.